fixtures/flight-esm/server/region.js

```diff
@@ -0,0 +1,140 @@
+'use strict';
+
+// This is a server to host data-local resources like databases and RSC
+
+const path = require('path');
+const url = require('url');
+
+if (typeof fetch === 'undefined') {
+  // Patch fetch for earlier Node versions.
+  global.fetch = require('undici').fetch;
+}
+
+const express = require('express');
+const bodyParser = require('body-parser');
+const busboy = require('busboy');
+const app = express();
+const compress = require('compression');
+const {Readable} = require('node:stream');
+
+app.use(compress());
+
+// Application
+
+const {readFile} = require('fs').promises;
+
+const React = require('react');
+
+const moduleBasePath = new URL('../src', url.pathToFileURL(__filename)).href;
+
+async function renderApp(res, returnValue) {
+  const {renderToPipeableStream} = await import('react-server-dom-esm/server');
+  const m = await import('../src/App.js');
+
+  const App = m.default;
+  const root = React.createElement(App);
+  // For client-invoked server actions we refresh the tree and return a return value.
+  const payload = returnValue ? {returnValue, root} : root;
+  const {pipe} = renderToPipeableStream(payload, moduleBasePath);
+  pipe(res);
+}
+
+app.get('/', async function (req, res) {
+  await renderApp(res, null);
+});
+
+app.post('/', bodyParser.text(), async function (req, res) {
+  const {
+    renderToPipeableStream,
+    decodeReply,
+    decodeReplyFromBusboy,
+    decodeAction,
+  } = await import('react-server-dom-esm/server');
+  const serverReference = req.get('rsc-action');
+  if (serverReference) {
+    // This is the client-side case
+    const [filepath, name] = serverReference.split('#');
+    const action = (await import(filepath))[name];
+    // Validate that this is actually a function we intended to expose and
+    // not the client trying to invoke arbitrary functions. In a real app,
+    // you'd have a manifest verifying this before even importing it.
```

Comment:
This would require a bundler, right? Just to confirm. Or does the loader do this somehow

Reply:
It's pretty simple because it'll have been tagged as a server reference by the loader. So verifying is trivial. The decodeReply/decodeAction also has this issue atm.

Reply:
The part I'm missing is, importing a file calls its top-level side effects, so isn't this a possible vector if the filename is arbitrary?